lib/block_view.go

```diff
@@ -2815,25 +2851,58 @@ func (bav *UtxoView) _disconnectNFTBid(
+	// These are "implicit" outputs that always occur at the end of the
+	// list of UtxoOperations. The number of implicit outputs is equal to
+	// the total number of "Add" operations minus the explicit outputs.
+	numUtxoAdds := 0
+	for _, utxoOp := range utxoOpsForTxn {
+		if utxoOp.Type == OperationTypeAddUtxo {
+			numUtxoAdds += 1
+		}
+	}
+	operationIndex -= numUtxoAdds - len(currentTxn.TxOutputs)
+
 	// Get the NFTBidEntry corresponding to this txn.
 	bidderPKID := bav.GetPKIDForPublicKey(currentTxn.PublicKey)
 	if bidderPKID == nil || bidderPKID.isDeleted {
 		return fmt.Errorf("_disconnectNFTBid: PKID for bidder public key %v doesn't exist; this should never happen", string(currentTxn.PublicKey))
 	}
-	nftBidKey := MakeNFTBidKey(bidderPKID.PKID, txMeta.NFTPostHash, txMeta.SerialNumber)
-	nftBidEntry := bav.GetNFTBidEntryForNFTBidKey(&nftBidKey)
-	// Sanity-check that it exists.
-	if nftBidEntry == nil || nftBidEntry.isDeleted {
-		return fmt.Errorf("_disconnectNFTBid: Bid entry for "+
-			"nftBidKey %v doesn't exist; this should never happen", nftBidKey)
-	}
+	// If there isn't a previous NFT entry, we're dealing with a bid on a non-BuyNow NFT.
+	if operationData.PrevNFTEntry == nil {
+
+		nftBidKey := MakeNFTBidKey(bidderPKID.PKID, txMeta.NFTPostHash, txMeta.SerialNumber)
+		nftBidEntry := bav.GetNFTBidEntryForNFTBidKey(&nftBidKey)
+		// Sanity-check that it exists.
+		if nftBidEntry == nil || nftBidEntry.isDeleted {
+			return fmt.Errorf("_disconnectNFTBid: Bid entry for "+
+				"nftBidKey %v doesn't exist; this should never happen", nftBidKey)
+		}
+
+		// Delete the existing NFT bid entry.
+		bav._deleteNFTBidEntryMappings(nftBidEntry)
 
-	// Delete the existing NFT bid entry.
-	bav._deleteNFTBidEntryMappings(nftBidEntry)
+		// If a previous entry exists, set it.
+		if operationData.PrevNFTBidEntry != nil {
+			bav._setNFTBidEntryMappings(operationData.PrevNFTBidEntry)
+		}
+	} else {
+		// If there is a previous NFT entry, that means the NFT entry was modified and that only happens for Buy Now NFTs when bidding.
+		if err := bav._helpDisconnectNFTSold(operationData, txMeta.NFTPostHash); err != nil {
+			return errors.Wrapf(err, "_disconnectNFTBid: ")
+		}
+		// Additionally, we need to delete the NFT bid since it did not exist before this transaction and was added back
+		// when we added back the DeletedNFTBidEntries
+		nftBidKey := MakeNFTBidKey(bidderPKID.PKID, txMeta.NFTPostHash, txMeta.SerialNumber)
+		nftBidEntry := bav.GetNFTBidEntryForNFTBidKey(&nftBidKey)
+		// Sanity-check that it exists.
+		if nftBidEntry == nil || nftBidEntry.isDeleted {
+			return fmt.Errorf("_disconnectNFTBid: Bid entry for "+
+				"nftBidKey %v doesn't exist; this should never happen", nftBidKey)
+		}
 
-	// If a previous entry exists, set it.
-	if operationData.PrevNFTBidEntry != nil {
-		bav._setNFTBidEntryMappings(operationData.PrevNFTBidEntry)
+		// Delete the existing NFT bid entry.
+		bav._deleteNFTBidEntryMappings(nftBidEntry)
 	}
```

Comment:
Feels like we can consolidate this block a bit:

```
if operationData.PrevNFTEntry != nil {
  if !operationData.PrevNFTEntry.IsBuyNow {
    // Error since this should never happen
  }
  // If we get here, we know this bid was on a buy-now NFT, and that the underlying NFT
  // was sold outright to the bidder.

  // If there is a previous NFT entry, that means the NFT entry was modified and that only happens for Buy Now NFTs when bidding.
  if err := bav._helpDisconnectNFTSold(operationData, txMeta.NFTPostHash); err != nil {
    return errors.Wrapf(err, "_disconnectNFTBid: ")
  }
}
// The steps after this point apply irrespective of whether the bid was on a buy-now NFT.

nftBidKey := MakeNFTBidKey(bidderPKID.PKID, txMeta.NFTPostHash, txMeta.SerialNumber)
nftBidEntry := bav.GetNFTBidEntryForNFTBidKey(&nftBidKey)
// Sanity-check that it exists.
if nftBidEntry == nil || nftBidEntry.isDeleted {
  return fmt.Errorf("_disconnectNFTBid: Bid entry for "+
    "nftBidKey %v doesn't exist; this should never happen", nftBidKey)
}
// Delete the existing NFT bid entry.
bav._deleteNFTBidEntryMappings(nftBidEntry)

// If a previous entry exists, set it. This will only be the case for non-buy-now NFTs.
if operationData.PrevNFTBidEntry != nil {
  // Sanity-check that there is no PrevNFTEntry
  if operationData.PrevNFTEntry != nil {
      // error
  }

  bav._setNFTBidEntryMappings(operationData.PrevNFTBidEntry)
}
```

Reply:
yeah this is smooth. thx!